test/test_optim.py

```diff
@@ -2199,6 +2200,202 @@ def test_non_empty_state(self, device, dtype, optim_info):
             for state in optim.state.values():
                 self.assertGreater(len(state), 0)
 
+    @optims(
+        [optim for optim in optim_db if optim.optim_cls.__name__ == "Adam"],
+        dtypes=[torch.float32],
+    )
+    def test_less_mem_beta1_zero_adam(self, device, dtype, optim_info):
+        # test that beta1=0.0 uses less memory than the default
+        model = torch.nn.Linear(5, 5)
+        model.to(dtype=dtype, device=device)
+        inpt = torch.rand(2, 5, dtype=dtype, device=device)
+
+        optim_inputs = optim_info.optim_inputs_func(device=device)
+        
+        mem_used_exp_avg = 0 
+        case_to_mem_usage = {}
+        
+        def get_obj_size(d):
+            if isinstance(d, dict):
+                return sum(get_obj_size(v) for v in d.values())
+            elif isinstance(d, (list,tuple)):
+                return sum(get_obj_size(v) for v in d)
+            elif isinstance(d, torch.Tensor):
+                return d.element_size() * d.numel()
+            return d.__sizeof__()
+            
+        for optim_input in optim_inputs:
+            beta1_zero = (
+                "betas" in optim_input.kwargs and optim_input.kwargs["betas"][0] == 0.0
+            )
+            if beta1_zero or optim_input.desc == "default":
+                optim = optim_info.optim_cls(model.parameters(), **optim_input.kwargs)
+                optim.zero_grad() 
+                output = model(inpt)
+                loss = output.sum()
+                loss.backward()
+                optim.step()
+
+                if not beta1_zero : 
```

Comment:
isn't this not going to be true since it's under the conditional in 2231?

Either way, this test can be simplified into the following:
```
    @optims(
        [optim for optim in optim_db if optim.optim_cls.__name__ == "Adam"],
        dtypes=[torch.float32],
    )
    def test_less_mem_beta1_zero_adam(self, device, dtype, optim_info):
        # test that beta1=0.0 uses less memory than the default
        beta_1_optim_inputs = [o for o in optim_info.optim_inputs_func(device=device) if "betas" in optim_input.kwargs and optim_input.kwargs["betas"][0] > 0.0]

        num_params = 4
        size_of_param_in_bytes = 32 * 16 * sizeof(dtype)  # hardcoded, also idk if sizeof is the right api)
        for optim_input in beta_1_optim_inputs:
            beta1_values = (optim_input.kwargs["betas"][0], 0.0)
            total_sizes = []    # will end up as [big_state_dict_size, no_exp_avg_sd_size]
            for beta1 in beta1_values:
                optim_input.kwargs["betas"] = (beta1, optim_input.kwargs["betas"][1])
                params = [torch.rand(32, 16, device=device, dtype=dtype) for _ in range(num_params)]
                for p in params:
                    p.grad = torch.rand_like(p)

                optim = optim_info.optim_cls(params, **optim_input.kwargs)
                for _ in range(3):
                    optim.step()
                    optim.zero_grad()
                    for p in params:
                        p.grad = torch.rand_like(p)

                total_sizes.append(get_obj_size(optim.state_dict()))

            # size of exp_avg should match the size of params, as all the params here requires grad
            size_of_exp_avgs = num_params * size_of_param_in_bytes
            self.assertEqual(total_sizes[0], total_sizes[1] + size_of_exp_avgs)